File: app/src/main/assets/modules/__toast__.js
// noinspection JSUnusedLocalSymbols,UnnecessaryLocalVariableJS

/**
 * @param {org.autojs.autojs.runtime.ScriptRuntime} scriptRuntime
 * @param {org.mozilla.javascript.Scriptable | global} scope
 * @return {Internal.Toast}
 */
module.exports = function (scriptRuntime, scope) {

    const ScriptToast = org.autojs.autojs.runtime.api.ScriptToast;

    let _ = {
        uiHandler: runtime.getUiHandler(),
        ToastCtor: (/* @IIFE */ () => {
            /**
             * @implements Internal.Toast
             */
            const ToastCtor = function () {
                // @Caution by SuperMonster003 on Oct 11, 2022.
                //  ! android.widget.Toast.makeText() doesn't work well on Android API Level 28 (Android 9) [P].
                //  ! There hasn't been a solution for this so far.
                //  ! Tested devices:
                //  ! 1. SONY XPERIA XZ1 Compact (G8441)
                //  ! 2. Android Studio AVD (Android 9.0 x86)
                /** @global */
                const toast = function (msg, isLong, isForcible) {
                    let $ = {
                        badge: {
                            long: /^l(ong)?$/i,
                            short: /^s(hort)?$/i,
                            forcible: /^f(orcible)?$/i,
                        },
                        toast() {
                            this.init(arguments);
                            this.show();
                        },
                        init() {
                            this.message = isNullish(msg) ? '' : String(msg);
                            this.isLong = this.parseIsLong(isLong);
                            this.isForcible = this.parseIsForcible(isForcible);
                        },
                        parseIsLong(o) {
                            let def = typeof this.isLong === 'boolean' ? this.isLong : false;
                            if (typeof o === 'boolean') {
                                return o;
                            }
                            if (typeof o === 'number') {
                                return Boolean(o);
                            }
                            if (typeof o === 'string') {
                                if (this.badge.long.test(o)) {
                                    return true;
                                }
                                if (this.badge.short.test(o)) {
                                    return false;
                                }
                                if (this.badge.forcible.test(o)) {
                                    this.isForcible = true;
                                    return def;
                                }
                                throw Error(`Invalid param: {name: isLong, value: ${o}, type: ${species(o)}.`);
                            }
                            return def;
                        },
                        parseIsForcible(o) {
                            let def = typeof this.isForcible === 'boolean' ? this.isForcible : false;
                            if (typeof o === 'boolean') {
                                return o;
                            }
                            if (typeof o === 'number') {
                                return Boolean(o);
                            }
                            if (typeof o === 'string') {
                                if (this.badge.forcible.test(o)) {
                                    return true;
                                }
                                if (this.badge.long.test(o)) {
                                    this.isLong = true;
                                    return def;
                                }
                                if (this.badge.short.test(o)) {
                                    this.isLong = false;
                                    return def;
                                }
                                throw Error(`Invalid param: {name: isForcible, value: ${o}, type: ${species(o)}.`);
                            }
                            return def;
                        },
                        show() {
                            if ($.isForcible) {
                                ToastCtor.prototype.dismissAll();
                            }
                            scriptRuntime.uiHandler.toast($.message, $.isLong);
                        },
                    };

                    $.toast();
                };

                return Object.assign(toast, ToastCtor.prototype);
            };

            ToastCtor.prototype = {
                constructor: ToastCtor,
                dismissAll() {
                    runtime.uiHandler.dismissAllToasts();
                },
            };

            return ToastCtor;
        })(),
    };

    /**
     * @type {Internal.Toast}
     */
    const toast = new _.ToastCtor();

    return toast;
};

File: app/src/main/java/org/autojs/autojs/runtime/api/augment/toast/Toast.kt
package org.autojs.autojs.runtime.api.augment.toast

import org.autojs.autojs.annotation.RhinoRuntimeFunctionInterface
import org.autojs.autojs.rhino.ArgumentGuards
import org.autojs.autojs.runtime.ScriptRuntime
import org.autojs.autojs.runtime.api.ScriptToast
import org.autojs.autojs.runtime.api.augment.Augmentable
import org.autojs.autojs.runtime.api.augment.Invokable
import org.autojs.autojs.runtime.exception.WrappedIllegalArgumentException
import org.autojs.autojs.util.RhinoUtils.UNDEFINED
import org.mozilla.javascript.Undefined

class Toast(private val scriptRuntime: ScriptRuntime) : Augmentable(scriptRuntime), Invokable {

    override val selfAssignmentFunctions = listOf(
        ::dismissAll.name,
    )

    // @Caution by SuperMonster003 on Oct 11, 2022.
    //  ! android.widget.Toast.makeText() doesn't work well on Android API Level 28 (9) [P].
    //  ! There hasn't been a solution for this so far.
    //  ! Tested devices:
    //  ! 1. SONY XPERIA XZ1 Compact (G8441)
    //  ! 2. Android Studio AVD (Android 9.0 x86)
    //  ! zh-CN:
    //  ! android.widget.Toast.makeText() 在安卓 API 级别 28 (9) [P] 上运行状况不佳.
    //  ! 到目前为止, 还没有针对这个问题的解决方案.
    //  ! 进行测试的设备包括:
    //  ! 1. 索尼 XPERIA XZ1 Compact (G8441 型号)
    //  ! 2. Android Studio 自带的安卓虚拟设备 (安卓 9.0 x86 架构)
    override fun invoke(vararg args: Any?): Undefined = call(scriptRuntime, args)

    companion object : ArgumentGuards() {

        @JvmStatic
        @RhinoRuntimeFunctionInterface
        fun call(scriptRuntime: ScriptRuntime, args: Array<out Any?>): Undefined = ensureArgumentsLengthInRange(args, 0..3) {
            val toastParser = when (it.size) {
                0 -> ToastParser(globalContext, "")
                1 -> ToastParser(globalContext, args[0])
                2 -> ToastParser(globalContext, args[0], args[1])
                3 -> ToastParser(globalContext, args[0], args[1], args[2])
                else -> throw WrappedIllegalArgumentException("Invalid arguments length ${args.size} for global.toast")
            }
            toastParser.show(scriptRuntime)
            return@ensureArgumentsLengthInRange UNDEFINED
        }

        @JvmStatic
        @RhinoRuntimeFunctionInterface
        fun dismissAll(scriptRuntime: ScriptRuntime, args: Array<out Any?>): Undefined = ensureArgumentsIsEmpty(args) {
            ScriptToast.dismissAll(scriptRuntime)
            return@ensureArgumentsIsEmpty UNDEFINED
        }

    }

}
